src/Storages/MergeTree/ReplicatedMergeTreePartCheckThread.cpp

```diff
@@ -555,17 +555,30 @@ void ReplicatedMergeTreePartCheckThread::run()
 
             selected = std::find_if(parts_queue.begin(), parts_queue.end(), [current_time](const auto & elem)
             {
-                return elem.second <= current_time;
+                return elem.time <= current_time;
             });
             if (selected == parts_queue.end())
+            {
+                // find next part to check in the queue and schedule the check
+                // otherwise, scheduled for later checks won't be executed until a new check is enqueued (i.e. task is scheduled again)
+                auto next_it =
+                    std::min_element(begin(parts_queue), end(parts_queue),
+                                     [](const auto &l, const auto &r) {
+                                       return l.time < r.time;
+                                     });
+                if (next_it != parts_queue.end()) {
+                    auto delay = next_it->time - current_time;
+                    task->scheduleAfter(delay * 1000);
```

Comment:
Isn't it equivalent to simply checking if the queue is empty? What's the point of searching the min element?

Reply:
No, it isn't. The code path reaches this place if no check was selected. So, here the queue either empty or contains element(s) with delayed check. So, `min_element` will find the nearest one (by time) to check and schedule it according to its desired time to execute. If we don't schedule the delayed check here, then the time for the delayed check to execute depends on if anyone will enqueue a part check.